File: old_plan/src/KenpinTool.Prototype/Services/CaseLoader.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using PdfiumViewer;

namespace KenpinTool.Prototype;

public sealed class CaseLoader
{
    private static readonly HashSet<string> ImageExtensions = new(StringComparer.OrdinalIgnoreCase)
    {
        ".jpg",
        ".jpeg",
        ".png",
        ".bmp",
        ".tif",
        ".tiff",
    };

    private static readonly HashSet<string> PdfExtensions = new(StringComparer.OrdinalIgnoreCase)
    {
        ".pdf",
    };

    public IReadOnlyList<PageSource> LoadPages(string folderPath)
    {
        if (string.IsNullOrWhiteSpace(folderPath))
        {
            throw new ArgumentException("Folder path is required.", nameof(folderPath));
        }

        if (!Directory.Exists(folderPath))
        {
            throw new DirectoryNotFoundException(folderPath);
        }

        var files = Directory.EnumerateFiles(folderPath, "*.*", SearchOption.TopDirectoryOnly)
            .Where(IsSupportedExtension)
            .OrderBy(path => path, StringComparer.OrdinalIgnoreCase);

        var pages = new List<PageSource>();

        foreach (var file in files)
        {
            var extension = Path.GetExtension(file);
            if (ImageExtensions.Contains(extension))
            {
                pages.Add(new PageSource(file, null));
                continue;
            }

            if (PdfExtensions.Contains(extension))
            {
                var pageCount = GetPdfPageCount(file);
                for (var i = 1; i <= pageCount; i++)
                {
                    pages.Add(new PageSource(file, i));
                }
            }
        }

        return pages;
    }

    private static bool IsSupportedExtension(string path)
    {
        var extension = Path.GetExtension(path);
        return ImageExtensions.Contains(extension) || PdfExtensions.Contains(extension);
    }

    private static int GetPdfPageCount(string filePath)
    {
        try
        {
            using var document = PdfDocument.Load(filePath);
            var count = document.PageCount;
            if (count <= 0)
            {
                throw new InvalidDataException("PDF has no pages.");
            }

            return count;
        }
        catch (Exception ex) when (ex is not InvalidDataException)
        {
            throw new InvalidDataException($"PDF load failed: {filePath}", ex);
        }
    }
}

public sealed record PageSource(string FilePath, int? PdfPageIndex);

File: old_plan/src/kenpintool/services/case_loader.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

from kenpintool.services.pdf_utils import get_pdf_page_count


SUPPORTED_IMAGE_EXTS = {".jpg", ".jpeg", ".png", ".tif", ".tiff", ".bmp"}
SUPPORTED_PDF_EXTS = {".pdf"}


@dataclass(frozen=True)
class PageSource:
    file_path: Path
    pdf_page_index: int | None = None


class CaseLoader:
    def load_pages(self, input_path: Path) -> list[PageSource]:
        pages: list[PageSource] = []
        for path in sorted(input_path.rglob("*")):
            if not path.is_file():
                continue
            ext = path.suffix.lower()
            if ext in SUPPORTED_IMAGE_EXTS:
                pages.append(PageSource(file_path=path))
            elif ext in SUPPORTED_PDF_EXTS:
                page_count = get_pdf_page_count(path)
                for idx in range(page_count):
                    pages.append(PageSource(file_path=path, pdf_page_index=idx))
        return pages
